### PR TITLE
Simplify chunked octree traversals

### PR DESCRIPTION
**Pull request by victorreijgwart**
# Description

Release 2.0.0 introduced experimental `NodePtr` and `NodeRef` classes, enabling traversal of chunked octrees as if they were regular octrees. This PR finalizes their implementation, uses them to simplify map operations and measurement integrators, and extends the `QueryAccelerator` to support chunked octrees. Benchmarks confirm these changes improve code clarity while slightly enhancing performance.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that causes existing functionality to not work as expected)
- [ ] Other (please describe):

## Detailed Summary

Prior to this PR, handling chunked octrees involved explicit reasoning about chunks, relative node positions, and parent-child chunk boundaries. This approach made the code harder to read, maintain, and debug.

Release 2.0.0 introduced `NodePtr` and `NodeRef`, inspired by STL’s `std::vector<bool>::bitref`, to emulate the semantics of standard octree nodes for chunked octrees. However, their implementation was experimental, and the performance impact compared to directly interacting with chunks was uncertain.

This PR:
1. Refines the semantics and implementation of `NodePtr` and `NodeRef`.
2. Rewrites chunked octree map operations and measurement integrators using these classes.
3. Extends the `QueryAccelerator` to support `HashedChunkedWaveletOctrees`.

Extensive benchmarks show that these changes not only simplify the codebase but also slightly improve performance.

In addition, several older code sections were cleaned up. While most changes are internal, notable updates affecting users are documented below.

## API Changes

List any changes to wavemap's APIs to help users update their code. Write "None" if there are no changes.

### C++ API:

* Additions

  * The `QueryAccelerator` now also supports `HashedChunkedWaveletOctrees`
  * It is now possible to erase individual nodes from chunked octrees using `ChunkedNdtree::NodeRef`

* Refactoring

  * Some type aliases (`using ... = ...`) nested in the Map classes have been moved or renamed

### Python API:

* Additions

  * Maps of type HashedChunkedWaveletOctree now also include fast batched accessors

### ROS1 Interface:

* Refactoring

  * The `TfTransformer` now returns the transform by value, and indicates lookup failures by returning std::nullopt

## Review Notes

@LionelOtt A general review would be great. Let me know if some parts of the code are difficult to follow or if the purpose of certain methods or classes is unclear. Then I'll add more comments and docstrings.

# Testing

### Automated Tests

The `NodePtr` and `NodeRef` classes are thoroughly tested via the `ndtree` data structure, map, and measurement integrator test suites. Additionally, the `QueryAccelerator` test suite has been extended to cover `QueryAccelerator<HashedChunkedWaveletOctree>`.

### Manual Tests

The changes were manually tested by rerunning the [Demos](https://ethz-asl.github.io/wavemap/pages/demos.html) on an AMD laptop and an Intel desktop.

### Benchmarks (To be completed by maintainers)

Newer College 20cm
| Model   | Version   |   RAM (MB) |   Map size (MB) |   CPU time (s) |   Wall time (s) |   AUC |
|:--------|:----------|-----------:|----------------:|---------------:|----------------:|------:|
| beam    | v2.1.2    |     153.31 |           13.80 |          92.30 |           41.38 |  0.91 |
| beam    | proposed  |     148.79 |           13.80 |          90.23 |           39.80 |  0.91 |
| ray     | v2.1.2    |     148.70 |           13.17 |          69.18 |           35.52 |  0.91 |
| ray     | proposed  |     151.21 |           13.07 |          67.86 |           34.75 |  0.91 |

Newer College 5cm
| Model   | Version   |   RAM (MB) |   Map size (MB) |   CPU time (s) |   Wall time (s) |   AUC |
|:--------|:----------|-----------:|----------------:|---------------:|----------------:|------:|
| beam    | v2.1.2    |     871.07 |          576.56 |        2384.99 |          194.63 |  0.97 |
| beam    | proposed  |     854.34 |          576.62 |        2369.48 |          191.39 |  0.97 |
| ray     | v2.1.2    |     826.55 |          409.39 |        1152.18 |          113.68 |  0.96 |
| ray     | proposed  |     818.27 |          412.49 |        1138.29 |          110.81 |  0.96 |

Pantopic Flat 5cm
| Model   | Version   |   RAM (MB) |   Map size (MB) |   CPU time (s) |   Wall time (s) |   AUC |
|:--------|:----------|-----------:|----------------:|---------------:|----------------:|------:|
| beam    | v2.1.2    |     122.03 |            6.44 |           6.42 |            3.59 |  0.99 |
| beam    | proposed  |     121.40 |            6.43 |           6.35 |            3.57 |  0.99 |
| ray     | v2.1.2    |     124.53 |            7.10 |           4.32 |            3.02 |  0.99 |
| ray     | proposed  |     123.55 |            7.18 |           4.14 |            2.90 |  0.99 |

Pantopic Flat 2cm
| Model   | Version   |   RAM (MB) |   Map size (MB) |   CPU time (s) |   Wall time (s) |   AUC |
|:--------|:----------|-----------:|----------------:|---------------:|----------------:|------:|
| beam    | v2.1.2    |     269.81 |           58.44 |          53.78 |            7.79 |  1.00 |
| beam    | proposed  |     269.70 |           58.33 |          53.79 |            7.64 |  1.00 |
| ray     | v2.1.2    |     261.01 |           61.94 |          25.02 |            4.92 |  1.00 |
| ray     | proposed  |     260.38 |           61.85 |          24.65 |            4.86 |  1.00 |

Note that we leave out the plots comparing the accuracy-over-distance of `proposed` and `v2.1.2`, as they fully overlap. The plots for `v2.1.2` can be seen in PR #83.

# Checklist

General

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated tests as required
- [x] Any required changes in dependencies have been committed and pushed

Documentation (where applicable)

- [x] I have updated the installation instructions (in docs/pages/installation)
- [x] I have updated the code's inline API documentation (e.g., docstrings)
- [x] I have updated the parameter documentation (in docs/pages/parameters)
- [x] I have updated/extended the tutorials (in docs/pages/tutorials)

### CODE REVIEW

File: library/python/src/maps.cc [hunk 1]
```diff
@@ -252,6 +252,68 @@ void add_map_bindings(nb::module_& m) {
...
+                 indices) {
+            // Create a query accelerator
+            QueryAccelerator<HashedChunkedWaveletOctree> query_accelerator{
+                self};
+            // Create nb::ndarray view for efficient access to the query indices
+            auto index_view = indices.view();
+            const auto num_queries = index_view.shape(0);
+            // Create the raw results array and wrap it in a Python capsule that
+            // deallocates it when all references to it expire
+            auto* results = new float[num_queries];
```
Comment (LionelOtt) [inline library/python/src/maps.cc:298]: I assume that's nanobind that wants to have raw memory fun rather than raii things?

**Comment by LionelOtt** [comment]
I read through it, seems like a lot of clean-up and unification and typing changes but overall it's not too complicated to follow in my view.

File: library/python/src/maps.cc (hunk 1 above)
Comment (victorreijgwart) [reply library/python/src/maps.cc:298]: Yes. I was also a bit surprised. Their docs seem to suggest this approach, and I couldn't find an RAII-based alternative.

**Review by LionelOtt — APPROVED** [review]
Changes look good.

**Comment by victorreijgwart** [comment]
Thanks @LionelOtt. I'll update the Changelogs and document the API changes, and then we can merge it.

**Pull request merged**